### ISSUE TITLE
Connection when both "Force Enryption" and "Extended Protection" are active

### ISSUE THREAD

**Issue by michaelenglert** [Microsoft/mssql-jdbc#963]
## Question
I'm trying to connect to a SQL Server (2012) using Windows Authentication where both `Force Enryption` is set to `true` and `Extended Protection` is set to `required`.

What I'm getting on the Client side is:
```
Login failed. The login is from an untrusted domain and cannot be used with Windows authentication. ClientConnectionId:<redacted>
```

Server side:
```
SSPI handshake failed with error code 0x80090346, state 46 while establishing a connection with integrated security; the connection has been closed. Reason: The Channel Bindings from this client are missing or do not match the established Transport Layer Security (TLS) Channel. The service might be under attack, or the data provider or client operating system might need to be upgraded to support Extended Protection. Closing the connection. Client's supplied SSPI channel bindings were incorrect.   [CLIENT: <redacted>]
```

I've tried many different configuration options (client side) including:
* `sslProtocol` all options
* `trustStore` alongside `trustStorePassword` with a self signed cert
* `trustServerCertificate`
* `serverSpn` alongside corresponding server setting
* `encrypt`

All results in the same error combination. OS instances are identical on Windows Sever 2012 R2 latest SP.

Any hints would be highly appreciated.

Thanks
Michael

**Comment by cheenamalhotra** [comment]
Hi @michaelenglert 

Please confirm if Client (_Windows Server 2012 R2 in your case_) has enabled Extended Protection support.

This blog will be worth looking into on how to enable it: 
https://blogs.technet.microsoft.com/srd/2009/12/08/extended-protection-for-authentication/

More details about Windows Update required for Extended Protection support: https://support.microsoft.com/en-us/help/973917/description-of-the-update-that-implements-extended-protection-for-auth

**Comment by michaelenglert** [comment]
Hi @cheenamalhotra,

confirmed as this is enabled by default in Windows Server 2012 R2.

**Comment by cheenamalhotra** [comment]
Just to put myself better on this, the error you are receiving "_The login is from an untrusted domain and cannot be used with Windows authentication_" is coming due to lack of configuration on client side.

From the link above, you may wanna check
- if extended protection is enabled by setting the value of the registry key HKLM\System\CurrentControlSet\Control\LSA\SuppressExtendedProtection to 0 on Client OS. 
- if SPN as provided in `serverSPN `connection property is registered in <extendedProtection> section of ApplicationHost.config file on server.

Since you are working with Kerberos negotiation, it is expected to be supported by default, further configuration for LmCompatibilityLevel may not be mandatory.

On the contrary, please also confirm:
- Connection succeeds on disabling Extended Protection on Server and Kerberos setup is done properly.
- What happens if a test application attempts connection from the same OS where SQL Server is present. 

Please also provide your connection string with repro code if possible in a standalone Java class.

**Comment by michaelenglert** [comment]
thanks for helping @cheenamalhotra 

Using this Connection String `jdbc:sqlserver://192.168.56.101:1433;serverSpn=MSSQLSvc/WIN-097Q29SOILR.corp.michi.com:1433;integratedSecurity=true;authenticationScheme=JavaKerberos;trustServerCertificate=true`
* `Force Enryption` is set to `false` --> **Successful connection**
* `Force Enryption` is set to `false` --> **Errors happen as seen above**

I've checked the Kerberos setup using the Kerberos config tool and also registered the registry key `HKLM\System\CurrentControlSet\Control\LSA\SuppressExtendedProtection` to `0`

**Comment by cheenamalhotra** [comment]
Hi @michaelenglert 

I hope you wanted to mention error occurs if **Force Encryption** is set to **true** and are you not passing `encrypt=true` in connection string?

**Comment by michaelenglert** [comment]
Hi @cheenamalhotra 

I'm sorry. I lost track of all the different setups. Same error happens also with `encrypt=true` in

**Comment by michaelenglert** [comment]
Hi @cheenamalhotra it would be great if you have another hint for me :-)

**Comment by cheenamalhotra** [comment]
Hi @michaelenglert 

I'll try to reproduce this in my environment, as I'm not entirely sure if this is a configuration problem or something else. Let me know if you have any pointers to problem root cause.

**Comment by michaelenglert** [comment]
that would be awesome @cheenamalhotra there is not much I did on the setup. Here's what I did on the Server side:
- Basic Windows Server 2012 R2 Installation
- Basic SQL Server 2012 Installation
- Installed Active Directory on the same host
- Configured Active Directory **User** to "Trust this User for delegation"
- Configured Active Directory **Computer** to "Trust this computer for delegation"
- Set `Extended Protection` for SQL Server to `Required`
- Set `Force Encryption` for SQL Server to `Yes` - **if set no `No` it works**
- Created a self signed certificate and configured the SQL Server to use it - **also tried without it**

On the Client Side I tried/used:
- `serverSpn=serverSpn`
- `integratedSecurity=true`
- `authenticationScheme=JavaKerberos`
- `trustServerCertificate=true`
- `encrypt=true`
- `trustStore=storeName`
- `
trustStorePassword=storePassword`

**Comment by cheenamalhotra** [comment]
Thanks @michaelenglert for the details.

Out of curiosity, have you tried to connect with any of the other SQL Server drivers?

**Comment by cheenamalhotra** [comment]
Hi @michaelenglert 

I configured Force Protocol Encryption + Extended Protection on SQL Server 2012 and am able to connect using Kerberos.

I could connect with below properties set using SQLServerDataSource.

```java
TrustStore trustStore = new TrustStore();
SQLServerDataSource ds = new SQLServerDataSource();
ds.setServerName("<your server name>");
ds.setTrustStore(trustStore.getFileName());
ds.setTrustStorePassword(TrustStore.TRUST_STORE_PASSWORD);
ds.setIntegratedSecurity(true);
try (Connection con = ds.getConnection()) {
     System.out.println("Connected");
}
```

Also, if it works without adding `ds.setAuthenticationScheme("JavaKerberos");` as default Negotiation is set to Kerberos.

**Comment by cheenamalhotra** [comment]
I followed same steps defined in https://blogs.technet.microsoft.com/srd/2009/12/08/extended-protection-for-authentication/ for configuring Client and Server. No extra configurations were needed.

**Comment by cheenamalhotra** [comment]
Ok, looks like I am able to reproduce the error now, but the issue isn't dependent on **Force Encryption** Turned On/Off but on **Extended Protection** when set to "Required" on SQL Server Instance configuration window.

Anyways, I'll investigate more into it and let you know soon!

**Comment by cheenamalhotra** [comment]
Hi @michaelenglert 

Upon investigations I discovered that in order to work with both **Extended Protection** and **Encryption** on Connection, client needs to supply Channel Binding which is accomplished by the client transferring the SPN and the CBT to the server in a tamperproof fashion. In absence of which server will not accept any requests from client. More details can be found [here](https://docs.microsoft.com/en-us/previous-versions/visualstudio/visual-studio-2008/dd639324(v=vs.90)).

As of now, the driver does not implement Channel Binding and thus this combination of **Extended Protection + Force Encryption** during connection is not supported. Turning one of them _OFF_ will result in a successful connection, as you must have figured out too.

**Comment by michaelenglert** [comment]
Hi @cheenamalhotra 

Thanks. I really appreciate that you took the time to investigate this!!

Does it make sense to convert this question to an enhancement request then?!

Thanks again
Michael

**Comment by michaelenglert** [comment]
thanks @cheenamalhotra

**Comment by salen2000** [comment]
Hi @cheenamalhotra 
I encountered the exactly problem.
I don't quite understand the statement you post above,
"As of now, the driver does not implement Channel Binding and thus this combination of Extended Protection + Force Encryption during connection is not supported."
My question is 
in order to make the connection support both "Extended Protection + Force Encryption", what should I do on the client side?

thank you in advance!

**Comment by rene-ye** [comment]
Hi @salen2000, you can't do anything on the client side as of now. Changes to the driver are needed to make this work.

**Comment by salen2000** [comment]
> Hi @salen2000, you can't do anything on the client side as of now. Changes to the driver are needed to make this work.

Thanks Rene,
I found some article mentioned the "Extended Protection" is **only** supported by "SQL Native Client" (i.e. one of the driver, that only support ODBC and OLE DB), however, because I am using JDBC (another driver) to connect to SQL server. 
Can I conclude, by using JDBC as of now, I cannot connect SQL Server with both "Extended Protection + Force Encryption" enabled?
This is the link about SQL Native Client
https://docs.microsoft.com/en-us/sql/relational-databases/native-client/sql-server-native-client?view=sql-server-2017
This is the link about Extended Protection only Support SQL Server Native Client
https://docs.microsoft.com/en-us/sql/tools/configuration-manager/protocols-for-mssqlserver-properties-advanced-tab?view=sql-server-2017

**Comment by cheenamalhotra** [comment]
Hi @salen2000 

Yes that is true for JDBC Driver, as I mentioned above.

**Comment by ricardoaviz** [comment]
Hello cheenamalhotra,
Thanks for all valuable contributions from you side on this issue.

Do we have anyone working on this issue?

This is stuck now for 8 months and I don’t see any solution or roadmap to solve this.

Is it possible to prioritize this?

Thanks!
With best regards,

Ricardo Aviz
Siemens

**Comment by David-Engel** [comment]
@ricardoaviz Adding channel binding support (and thus, extended protection support) to the JDBC driver is on our future roadmap, but we do not have a specific timeline for it at this time. I've bumped it up the list a bit and I do think it's an important security feature but we'll have to see where it makes the cut when planning future releases.

**Comment by vinaykariyappa1** [comment]
com.microsoft.sqlserver.jdbc.SQLServerException: Login failed. The login is from an untrusted domain and cannot be used with Windows authentication. ClientConnectionId:dd5204e9-85ad-4625-be53-758b67062a5c
	at com.microsoft.sqlserver.jdbc.SQLServerException.makeFromDatabaseError(SQLServerException.java:262)

Is this issue is resolved in any other new version of JDBC driver...

**Comment by ricardoaviz** [comment]
Hi Vinay,

What is your post about?

Can you please clarify your post?

Tks

**Comment by vinaykariyappa1** [comment]
Jre version we are using 1.8
Jdbc we are using 7.2
jdbc:sqlserver://;servername=server_name;integratedSecurity=true; 
placed sqljdbc_auth.dll to C:\Program Files\Java\jre\bin and C:\Program Files\Java\jdk\bin folder. 
getting below error "This Driver is not configured for integrated authentication". kindly help to resolve this issue. 
will JDBC driver support integratedSecurity????

**Comment by ulvii** [comment]
Hi @vinaykariyappa1 ,
Driver already supports Windows integrated authentication, try setting `-Djava.library.path=path_to_DLL` in your VM arguments.

**Comment by gstanchev** [comment]
Hi, we've ran into same situation. Any expectations when channel binding support would be added to the driver?

**Comment by peterbae** [comment]
It's in the backlog, but we don't have plans on adding this to the driver for now. We will look into working on this in the next upcoming semester.

**Comment by postfach33** [comment]
Hello @peterbae .
Any updates on this issue?

**Comment by postfach33** [comment]
> @ricardoaviz Adding channel binding support (and thus, extended protection support) to the JDBC driver is on our future roadmap, but we do not have a specific timeline for it at this time. I've bumped it up the list a bit and I do think it's an important security feature but we'll have to see where it makes the cut when planning future releases.

Hello @David-Engel . are there any news to this topic?

**Comment by postfach33** [comment]
> 
> Upon investigations I discovered that in order to work with both **Extended Protection** and **Encryption** on Connection, client needs to supply Channel Binding which is accomplished by the client transferring the SPN and the CBT to the server in a tamperproof fashion. In absence of which server will not accept any requests from client. More details can be found [here](https://docs.microsoft.com/en-us/previous-versions/visualstudio/visual-studio-2008/dd639324(v=vs.90)).
> 
> As of now, the driver does not implement Channel Binding and thus this combination of **Extended Protection + Force Encryption** during connection is not supported. Turning one of them _OFF_ will result in a successful connection, as you must have figured out too.

Hi @cheenamalhotra ,
it seems as if this issue was removed from the ToDo List. I can't find it on the Board-->https://github.com/microsoft/mssql-jdbc/projects/2  Maybe I look at the wrong one?

**Comment by David-Engel** [comment]
@postfach33 There are no updates to share at this time. This issue is still in the backlog.

**Comment by mariush17** [comment]
@David-Engel and @peterbae Is there any timeline to implement the channel binding feature into the JDBC driver? We are working with Java in a Linux environment and need to access SQL Server databases with Extended Protection and Forced Encryption as this is a requirement from our organization. The implementation of the channel binding feature would help us a lot!

**Comment by David-Engel** [comment]
@mariush17 I'm sorry, there is no current timeline for this feature in the JDBC driver.

**Comment by juantarquino-ssb** [comment]
@David-Engel any news on adding support for channel binding feature into the JDBC driver?

**Comment by v-zhangw** [comment]
@juantarquino-ssb 
Sorry but no update as of yet.

**Comment by juantarquino-ssb** [comment]
Is there any workaround for using integrated authentication with servers that have those two things enabled?  or an alternative driver that does support channel binding for sql server? This issue is preventing us from using Java internally to connect to sql servers because we need to use integrated authentication in Windows.

**Comment by lilgreenbird** [comment]
Hi @juantarquino-ssb 

Channel binding support is on our radar, however unfortunately it did not make the cut for this semester, we will look will into this again when we plan for the next semester.  

fyi, we do take into account user requests when we do planning, so if there is any feature you would like implemented we encourage you do fill out our [survey](https://forms.office.com/Pages/ResponsePage.aspx?id=DQSIkWdsW0yxEjajBLZtrQAAAAAAAAAAAAFhcKCP4bxUQUxBNDA1NjRER0Y5SVhRQ1hSN0g1Q09MQS4u).  Thanks.

**Comment by Neustradamus** [comment]
@ all: It is official for TLS 1.3 Binding!
- RFC 9266: Channel Bindings for TLS 1.3: https://tools.ietf.org/html/rfc9266

Details:
- tls-unique for TLS =< 1.2 (RFC5929)
- tls-server-end-point =< 1.2 + 1.3 (RFC5929)
- tls-exporter for TLS = 1.3 (RFC9266)

Thanks in advance.

**Comment by dineubr** [comment]
> Just an update for those following the thread, we are still in the middle of investigating this.
> 
> From my initial investigation, to enable channel bindings in the driver I applied the following fix to the `intAuthHandShake` to `KerbAuthentication.java` file:
> 
> ```
> InetAddress initiator = InetAddress.getLocalHost();
> InetAddress acceptor = InetAddress.getByName(con.activeConnectionProperties
>                     .getProperty(SQLServerDriverStringProperty.SERVER_NAME.toString()));
> 
> ChannelBinding cb = new ChannelBinding(initiator, acceptor, null);
> peerContext.setChannelBinding(cb);
> ```
> 
> However, the above change doesn't seem to have an impact and I still receive the login error. I'll continually post updates as I investigate.
> 
> ### Updates (I'll post brief updates in this post to avoid spam)
> * Aug 17th 2022 - Issue is still on my radar, investigating the difference between the JDBC driver and other drivers to narrow down cause

Hi @tkyc, 
Would you have any updates regarding this investigation?

Regards,
Dineu

**Comment by Jeffery-Wasty** [comment]
Hi @dineubr,

We're still working on this. Terry released an update in an edit to his most recent comment:

> ### Updates (I'll post brief updates in this post to avoid spam)
> * Aug 17th 2022 - Issue is still on my radar, investigating the difference between the JDBC driver and other drivers to narrow down cause

**Comment by dineubr** [comment]
> Hi @dineubr,
> 
> We're still working on this. Terry released an update in an edit to his most recent comment:
> 
> > ### Updates (I'll post brief updates in this post to avoid spam)
> > 
> > * Aug 17th 2022 - Issue is still on my radar, investigating the difference between the JDBC driver and other drivers to narrow down cause

Hi @Jeffery-Wasty  and @tkyc,

Do we have any updates regarding this investigation?

Regards,
Dineu

**Comment by user1oc** [comment]
Hi are there any updates on this issue?
As per the last comment from @David-Engel and @tkyc, this feature was put on hold.

**Comment by David-Engel** [comment]
@user1oc No new updates. Lack of appropriate public Java APIs is still a blocker for this issue. I don't see that changing any time soon.

**Comment by ecki** [comment]
BTW it’s also an issue for http and ldap and really needs reconsideration on openjdk level. Maybe escalate it to the Microsoft Java team?

**Comment by tkyc** [comment]
@ecki Yeah, we've considered that as an option, but for now even with support from the Microsoft Java team I don't see the openjdk team consenting to the required Java API changes to make this happen as these types of changes were requested in the past.

**Comment by ecki** [comment]
> Force Encryption: On; Extended Protection: Allowed --> Same error ...why?

@martin-guth thats indeed problematic, I suspect the SSPI on the client think it can/should negotiate it and then Java does not add the token therefore corrupting the handshake which the server uses to abort. Would maybe a good idea to filter it out in the driver so it does not get negotiated.

**Comment by postfach33** [comment]
I would actually relabel the Issue from 'Enhancement' to 'Security Issue' as we are forced to downgrade the complete Instance, because if just 1 Client still can't use Channel binding nowadays, all other clients have to use the downgraded security as well.

**Comment by ADror-Quest** [comment]
any update on when this will be fixed?

**Comment by ecki** [comment]
> Why is Extended Protection: Allowed not possible? AFAIK this is the setting at the server side which does not enforce Extended Protection.

I think the drives should not claim it Supports it and then fail to provide Proof (due to JSSE Restriktion).

**Comment by lilgreenbird** [comment]
hi @ADror-Quest 

Unfortunately we don't have a timeline to address this as this feature is still currently blocked by a lack of TLS APIs in Java required to implement it. 

We did work on this in a feature branch where we worked around this using reflection ([channel-bindings](https://github.com/microsoft/mssql-jdbc/tree/channel-bindings)) or Bouncy Castle ([channel-bindings-bc](https://github.com/microsoft/mssql-jdbc/tree/channel-bindings-bc)) . These are **not** official solutions they will not be merged however you are welcome to take a look and build a version of the driver for your needs but please be aware this is **not** something that we officially support. Until there is a proper API available in Java this will not be something that will be supported in the driver.

**Comment by postfach33** [comment]
any update on when this will be fixed?
